python/shotgun_model/data_handler.py

```diff
@@ -0,0 +1,473 @@
+import os
+import datetime
+import cPickle
+import time
+
+# toolkit imports
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+from .errors import ShotgunModelDataError
+
+
+def log_timing(func):
+    """
+    Decorator that times and logs the execution of a method.
+    Borrowed from 0.18.
+    """
+    def wrapper(self, *args, **kwargs):
+        time_before = time.time()
+        try:
+            response = func(self, *args, **kwargs)
+        finally:
+            time_spent = time.time() - time_before
+            # log to special timing logger
+            self._bundle.log_debug("ShotgunDataHandler.%s took %fs" % (func.__name__, time_spent))
+        return response
+    return wrapper
+
+
+class ShotgunDataHandler(QtCore.QObject):
+    """
+    Abstract class that manages low level data storage for QT models.
+
+    This class abstracts away the data management and allows
+    the model to access the data in a simple tree-like fashion.
+    Each node in the tree is also identified by a unique id
+    and can be accessed directly via this id in an O(1) lookup.
+
+    It also offers fast serialization and loading. Each
+    ShotgunDataHandler is connected to a single cache file on disk.
+
+    Each QT model typically sublcasses this in order to create
+    a specific data handler. These methods need to be implemented
+    by all deriving classes:
+
+    - generate_data_request - called by the model when it needs
+      additional data to be loaded from shotgun. The data handler
+      formulates the exact request to be sent out to the server.
+
+    - update_data - the counterpart of generate_data_request: this
+      is called when the requested shotgun data is returned and
+      needs to be inserted into the data structure.
+
+    Data returned back from this class to the Model layer
+    is always sent as ShotgunDataItem object to provide a full
+    encapsulation around the internals of this class.
+    """
+    # version of binary format - increment this whenever changes
+    # are made which renders the cache files non-backwards compatible.
+    FORMAT_VERSION = 27
+
+    # constants for updates
+    (UPDATED, ADDED, DELETED) = range(3)
+
+    # internal constants for serialization performance
+    (CACHE_BY_UID, CACHE_CHILDREN, IS_LEAF, UID, PARENT, FIELD, SG_DATA) = range(7)
+
+    def __init__(self, cache_path, parent):
+        """
+        :param cache_path: Path to cache file location
+        :param parent: Parent QT object
+        """
+        super(ShotgunDataHandler, self).__init__(parent)
+        # keep a handle to the current app/engine/fw bundle for convenience
+        self._bundle = sgtk.platform.current_bundle()
+        # the path to the cache file
+        self._cache_path = cache_path
+        # data in cache
+        self._cache = None
+
+    def __repr__(self):
+        """
+        Create a string representation of this instance
+        :returns: A string representation of this instance
```

Comment:
Hah! I think you actually wrote this code 😆  I agree i think it's OTT. I'll dial it down 👻